test/Microsoft.ML.Benchmarks/PredictionEngineBench.cs

```diff
@@ -127,7 +127,7 @@ public void SetupBreastCancerPipeline()
             IDataView data = loader.Load(_breastCancerDataPath);
 
             var pipeline = env.BinaryClassification.Trainers.SdcaNonCalibrated(
-                new SdcaNonCalibratedBinaryTrainer.Options { NumberOfThreads = 1, ConvergenceTolerance = 1e-2f, });
+                new SdcaNonCalibratedBinaryTrainer.Options { NumberOfThreads = 1, ConvergenceTolerance = 1e-2f, Shuffle = false});
```

Comment:
It seems we better create different tests to test againt Shuffle=true and Shuffle=false.
  1. when Shuffle = false, the result should be deterministic and we can assert value
  2. when Shuffle = true, the result should lie in certain range and we can assert range

Reply:
With `shuffle = false` you won't be fully deterministic as there's always race conditions in due to threading.

Easy example of a four core machine with four threads training a model slightly differently than a two core machine. 

Lets assume there's either (1) a single shared PRNG in a component where N threads pull random numbers at a slightly different ordering each run, or (2) each thread has its own PRNG where the number of threads affect the random numbers generated.

Reply:
I see, thanks for your explain.
So in that case, we can use a way (for example mock the parts that is not deterministic) to gain deterministic and be able to do test specific part of code is correct, this is what we usually do for unit test (mock operation that is not deterministic or have outer dependency).
In integration test, we can assert on range to ensure whole scenario is correct.


---
In reply to: [373736059](https://github.com/dotnet/machinelearning/pull/4750#discussion_r373736059) [](ancestors = 373736059)